### PR TITLE
Flux aux: Compute flux in diffusion problems. 

### LINKED ISSUE #10050: Need generic AuxKernel to compute flux for diffusion problems
## Rationale
The MOOSE framework provides a basic Diffusion kernel, and diffusion problems are solved using either that or derived classes in a lot of applications. I would like to have the ability to output the components of the flux vector for diffusion problems. 

## Description
This seems analogous to the diffusion kernel, so I think it belongs in the framework. The flux is computed as $$J=-D\frac{\partial C}{\partial X}$$, where J is the flux, D is the diffusivity, C is the concentration variable, and X is the coordinate. 

## Impact
Add an AuxKernel (I'd name it `FluxAux`) that computes components of the flux vector using the gradient of the diffusion variable and the diffusivity. Also add a test in moose/test/tests/auxkernels.

### PR DESCRIPTION
**Pull request by yeh512**
closes #10050

### CODE REVIEW

**Comment by bwspenc** [comment]
I'm working with @yeh512 on this, and will do the review.

**Comment by bwspenc** [comment]
It's failing on the Precheck.  If you click to see the details, you'll see that you have these errors that need to be corrected:
```
##########################################################################
ERROR: Your patch does not contain a valid ticket reference! (i.e. #1234)
Merge branch 'flux_aux' of github.com:yeh512/moose into test
add test output file, ref. issue # 10050
Add tests for FluxAux, ref issue # 10050
Add an auxkernel to compute flux in diffusion problems, ref issue # 10050

ERROR: The following files contain trailing whitespace after applying your patch:
	test/tests/auxkernels/diffusion_flux/diffusion_flux.i

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.
##########################################################################
```

You referenced the issue in your commit messages, but you just formatted it wrong.  You need to remove the word 'issue' (ref issue #10050 -> ref #10050).

File: framework/include/auxkernels/FluxAux.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
+
+#ifndef FLUXAUX_H
+#define FLUXAUX_H
+
```
Comment (bwspenc) [inline framework/include/auxkernels/FluxAux.h]: You'll need the MOOSE headers at the top of all source and header files.  Copy from another file in this directory.

File: framework/src/auxkernels/FluxAux.C [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#include "FluxAux.h"
```
Comment (bwspenc) [inline framework/src/auxkernels/FluxAux.C]: You'll need the MOOSE headers at the top of all source and header files. Copy from another file in this directory.

File: framework/include/auxkernels/FluxAux.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+{
+public:
+  FluxAux(const InputParameters & parameters);
+
+protected:
+  /**
+   * AuxKernels MUST override computeValue.  computeValue() is called on
+   * every quadrature point.  For Nodal Auxiliary variables those quadrature
+   * points coincide with the nodes.
+   */
```
Comment (bwspenc) [inline framework/include/auxkernels/FluxAux.h]: Remove the comment for this function.  Not needed for override functions.

File: framework/include/auxkernels/FluxAux.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+#include "AuxKernel.h"
+
+// Forward Declarations
+class FluxAux;
+
+template <>
+InputParameters validParams<FluxAux>();
+
+/**
+ * Auxiliary kernel responsible for computing the flux in diffusion problems
```
Comment (bwspenc) [inline framework/include/auxkernels/FluxAux.h]: Maybe note that it is for components of the flux vector.

File: framework/src/auxkernels/FluxAux.C [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+template <>
+InputParameters
+validParams<FluxAux>()
+{
+  InputParameters params = validParams<AuxKernel>();
+  // Declare the options for a MooseEnum.
+  // These options will be presented to the user in Peacock
+  // and if something other than these options is in the input file
+  // an error will be printed
```
Comment (bwspenc) [inline framework/src/auxkernels/FluxAux.C]: I'd delete these four lines of comments. They're not particularly helpful.

File: framework/src/auxkernels/FluxAux.C [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  MooseEnum component("x y z");
+
+  // Use the MooseEnum to add a parameter called "component"
```
Comment (bwspenc) [inline framework/src/auxkernels/FluxAux.C]: I don't think this comment is helpful either. Just delete it.

File: framework/src/auxkernels/FluxAux.C [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+                               "The name of the variable");
+  params.addRequiredParam<MaterialPropertyName>(
+      "diffusivity",
+      "The name of the diffusivity material property that will be used in the flux computation.");
+  return params;
+}
+
+  FluxAux::FluxAux(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    // This will automatically convert the MooseEnum to an integer
```
Comment (bwspenc) [inline framework/src/auxkernels/FluxAux.C]: Delete the comments and spaces in the construction list.

File: framework/src/auxkernels/FluxAux.C [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  MooseEnum component("x y z");
+
+  // Use the MooseEnum to add a parameter called "component"
+  params.addRequiredParam<MooseEnum>("component", component, "The desired component of flux.");
+
+  params.addRequiredCoupledVar("diffusion_variable",
+                               "The name of the variable");
+  params.addRequiredParam<MaterialPropertyName>(
+      "diffusivity",
+      "The name of the diffusivity material property that will be used in the flux computation.");
```
Comment (bwspenc) [inline framework/src/auxkernels/FluxAux.C]: Use `params.addClassDescription()` to add a short description of the class. 
You'll also need to add a Markdown file in moose/docs/content/documentation/systems/AuxKernels/framework to document this class. Use one of the existing ones in that directory as a template.

**Review by bwspenc — CHANGES_REQUESTED** [review]
The code looks good.  Just cosmetic stuff to fix.

File: docs/content/documentation/systems/AuxKernels/framework/FluxAux.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# FluxAux
+!syntax description /AuxKernels/FluxAux
+
+## Description
+The `FluxAux` AuxKernel is used to compute the components of the flux vector of Fick's Law. The flux is computed as $$J=-D\frac{\partial C}{\partial X}$$, where J is the diffusion flux vector, D is the diffusivity or diffusion coefficient, C is the concentration variable, and X is the coordinate.
```
Comment (bwspenc) [inline docs/content/documentation/systems/AuxKernels/framework/FluxAux.md]: components of the flux vector for a diffusion problems.

File: docs/content/documentation/systems/AuxKernels/framework/FluxAux.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+## Example Input Syntax
+!listing test/tests/auxkernels/diffusion_flux.i 
+
+!syntax parameters /AuxKernels/FluxAux
+
+!syntax inputs /AuxKernels/FluxAux
+   components
```
Comment (bwspenc) [inline docs/content/documentation/systems/AuxKernels/framework/FluxAux.md]: Delete these three lines: components, difusion variable, diffusivity:...

File: framework/src/auxkernels/FluxAux.C [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+/****************************************************************/
+#include "FluxAux.h"
+
+template <>
+InputParameters
+validParams<FluxAux>()
+{
+  InputParameters params = validParams<AuxKernel>();
+  MooseEnum component("x y z");
+  params.addClassDescription("Computing diffusive flux vector by Fick's Law "
```
Comment (bwspenc) [inline framework/src/auxkernels/FluxAux.C]: Compute components of flux vector for diffusion problems

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: test/tests/auxkernels/diffusion_flux/diffusion_flux.i [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+  [../]
+  [./flux_y]
+      order = FIRST
+      family = MONOMIAL
+  [../]
+[]
+
+[Kernels]
+  [./diff]
+    type = HeatConduction # A Laplacian operator
```
Comment (bwspenc) [inline test/tests/auxkernels/diffusion_flux/diffusion_flux.i]: You need to change this to use a Diffusion kernel

Comment (bwspenc) [reply test/tests/auxkernels/diffusion_flux/diffusion_flux.i]: Actually, use the MatDiffusion kernel, with the parameter 'prop_name = thermal_conductivity'.

**Comment by bwspenc** [comment]
@yeh512 It's failing the tests because the gold file for your test is missing.  You checked it in to the location:
test/tests/auxkernels/diffusion_flux/diffusion_flux_out.e
but it needs to be in:
test/tests/auxkernels/diffusion_flux/gold/diffusion_flux_out.e
If you go to that diffusion_flux directory, just do the following:
```
mkdir gold #assuming the directory doesn't already exist
git mv diffusion_flux_out.e gold/diffusion_flux_out.e
git commit --amend
```

**Comment by bwspenc** [comment]
@yeh512 See if you can get it to converge to a tighter tolerance. Try setting nl_rel_tol to something like 1e-12.  Keep tightening it up until it won't converge anymore, and then set it to something like 100 times that, and update the gold file.  That should fix the test diff problem.

Also, I talked with @YaqiWang about the naming of this, and we agreed that we should change the name of this to `DiffusionFluxAux` to make it clear that this is the flux associated with a diffusion equation. Please change the names of the files accordingly.

**Comment by YaqiWang** [comment]
Thanks for the naming consideration. We are living in a multiphysics world ;-)

File: test/tests/auxkernels/diffusion_flux/tests [hunk 13]
```diff
@@ -0,0 +1,7 @@
+[Tests]
+  [./test]
+    type = 'Exodiff'
+    input = 'diffusion_flux.i'
+    exodiff = 'diffusion_flux_out.e'
```
Comment (bwspenc) [inline test/tests/auxkernels/diffusion_flux/tests]: Add this line to get it to pass everywhere:
```
abs_zero = 1e-9
```

Comment (yeh512) [reply test/tests/auxkernels/diffusion_flux/tests]: do you mean   nl_abs_tol = 1e-9?

Comment (bwspenc) [reply test/tests/auxkernels/diffusion_flux/tests]: No.  Add that abs_zero to this file.  It sets the 'floor' used in the exodiff comparison, which makes it skip the check if a variable has a value below that floor.

**Review by bwspenc — APPROVED** [review]

**Comment by bwspenc** [comment]
I'm good with these changes.  Could someone on the MOOSE team (maybe @permcody) take a quick look at this since it adds framework code?

**Comment by permcody** [comment]
Looks ok

**Pull request merged**